lib/em-hiredis/state_machine.rb

```diff
@@ -0,0 +1,36 @@
+require 'set'
+
+module EventMachine::Hiredis
+  class StateMachine
+    include EventMachine::Hiredis::EventEmitter
+
+    attr_reader :state
+
+    def initialize
+      @transitions = {}
+      @state = :initial
+      @all_states = Set.new([:initial])
+    end
+
+    def add_transition(name, from, to)
+      existing = @transitions.find { |_, v| v == [from, to] }
+      if existing
+        raise "Duplicate transition #{from}, #{to}, already exists as #{existing.first}"
+      end
+
+      @all_states.add(from)
+      @all_states.add(to)
+      @transitions[name] = [from, to]
+    end
+
+    def update_state(to)
```

Comment:
Potentially any exceeding argument could be passed to the event ? (eg: the exception in the error transition)